java/arcs/sdk/testing/BaseTestHarness.kt

```diff
@@ -110,14 +120,36 @@ open class BaseTestHarness<P : Particle>(
      *
      * TODO: Describe the boot up sequence in detail once it is finalized.
      */
-    suspend fun start() {
+    suspend fun start() = coroutineScope {
         assertWithMessage("Harness can be started only once")
             .that(::particle.isInitialized).isFalse()
         particle = factory(scope)
         handles.forEach { (name, handle) -> particle.handles.setHandle(name, handle) }
+
         particle.onCreate()
-        handles.values.toList().withIndex().forEach { (index, handle) ->
-            particle.onHandleSync(handle, index == handles.size - 1)
+
+        val readySoFar = atomic(0)
+        val readyJobs = handles.map { (name, handle) ->
+            launch {
+                // TODO: switch to this version when onReady is non-suspending:
+                // suspendCoroutine<Unit> { cont ->
+                //    handle.onReady {
+                //        val ready = readySoFar.incrementAndGet()
+                //        cont.resume(Unit)
+                //        launch { particle.onHandleSync(handle, ready == handles.size) }
+                //    }
+                // }
+                // TODO: get rid of this version when onReady is non-suspending
+                val job = Job()
+                handle.onReady {
+                    val ready = readySoFar.incrementAndGet()
+                    job.complete()
+                    launch { particle.onHandleSync(handle, ready == handles.size) }
+                }
+                withTimeout(1000) { job.join() }
```

Comment:
How the 1000 was determined? If a timeout really happens and an exception is thrown, the enclosing coroutineScope will be cancelled completely. Is it intended?

Reply:
Yeah, if a timeout happens here, the whole world has broken - so it makes sense to fail the test.

Reply:
The 1000 was just a ballpark for now - we'll get rid of it alltogether once I get the final stuff submitted and these todos resolved.

Reply:
SGTM.